### PR TITLE
Show placeholders for images while loading, and then fade in images

### PR DESCRIPTION
**Pull request by perclasson**
Add placeholders for when images are loading. To make sure that the placeholders are the same size as the images, I added the image aspect ratio to the data of the demos. I assume in a _real application_ such information would be saved in a database.

fade_in_final_2

I am still waiting on the final designs for the placeholders. I'll update them in case there are any updates.

Closes https://github.com/material-components/material-components-flutter-gallery/issues/552

### CODE REVIEW

**Comment by guidezpl** [comment]
Is it possible to get around the size shift going from the placeholder to the image? Other than that, looks great!

**Comment by perclasson** [comment]
> Is it possible to get around the size shift going from the placeholder to the image? Other than that, looks great!

The shifting comes from the fact that the height of the placeholders are different compared to the images. We will only know the height of the images when they have been loaded. So I think it will be hard to get around the size shift going on.

**Comment by willlarche** [comment]
Looks much better! Can you do the same for the Shrine logo?

**Comment by perclasson** [comment]
I udpated the PR now with aspect ratios for the images to make sure that the placeholders have the exact same size as the images.
> Looks much better! Can you do the same for the Shrine logo?

Sure. I added it to the logos now.

**Comment by willlarche** [comment]
Can you put in the aspect ratio for the shrine logo?

**Comment by willlarche** [comment]
Is this necessary on mobile?

**Comment by perclasson** [comment]
> Can you put in the aspect ratio for the shrine logo?

Good idea.

> Is this necessary on mobile?

No. It should only be needed for the web. Let me update my FadeImagePlaceholder to only have this on the web and then to use this everywhere instead of FadeImage.

File: lib/layout/image_placeholder.dart [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import 'package:flutter/material.dart';
+
+/// An image that shows a [placeholder] widget while the target [image] is
+/// loading, then fades in the new image when it loads.
+///
+/// This is similar to [FadeInImage] but the difference is that it allows you
+/// to specify a widget as a [placeholder], instead of just an [ImageProvider].
+/// It also lets you override the [child] argument, in case you want to wrap
+/// the image with another widget, for example an [Ink.image].
+class FadeInImagePlaceholder extends StatelessWidget {
```
[the remark this answers is not in the document]
Comment (perclasson) [reply lib/layout/image_placeholder.dart]: That is already a widget inside the Flutter framework, and this one works a bit different.

What do you think about:
1. `WebFadeInImage`
2. `FadeInImageWithWidgetsPlaceholder`
3. `FadeInLoadingImage`

**Pull request merged**